test/cadet/devices/devices_test.exs

```diff
@@ -0,0 +1,228 @@
+                 thing_name: _
+               }
+             } =
+               Devices.get_device_ws_endpoint(device.id, user,
+                 datetime: {{2020, 01, 01}, {12, 34, 56}}
+               )
+    end
+  end
+
+  test "get ws endpoint by device secret", %{device: device, user: user} do
+    use_cassette "aws/devices_get_ws_endpoint#1", custom: true do
+      assert {
+               :ok,
+               %{
+                 client_name_prefix: _,
+                 endpoint:
+                   "wss://test-ats.iot.ap-southeast-1.amazonaws.com/mqtt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIATESTTEST%2F20200101%2Fap-southeast-1%2Fiotdevicegateway%2Faws4_request&X-Amz-Date=20200101T123456Z&X-Amz-Expires=300&X-Amz-SignedHeaders=host&X-Amz-Signature=49cc00fa6d52ad194bc9d63997bbbda4f6cac9c3f838ec3970fed65748550940&X-Amz-Security-Token=IOTESTTESTTEST",
+                 thing_name: _
+               }
+             } =
+               Devices.get_device_ws_endpoint(device.secret, user,
+                 datetime: {{2020, 01, 01}, {12, 34, 56}}
+               )
+    end
+  end
+
+  test "get ws endpoint by device", %{device: device, user: user} do
+    use_cassette "aws/devices_get_ws_endpoint#1", custom: true do
+      assert {
+               :ok,
+               %{
+                 client_name_prefix: _,
+                 endpoint:
+                   "wss://test-ats.iot.ap-southeast-1.amazonaws.com/mqtt?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=ASIATESTTEST%2F20200101%2Fap-southeast-1%2Fiotdevicegateway%2Faws4_request&X-Amz-Date=20200101T123456Z&X-Amz-Expires=300&X-Amz-SignedHeaders=host&X-Amz-Signature=49cc00fa6d52ad194bc9d63997bbbda4f6cac9c3f838ec3970fed65748550940&X-Amz-Security-Token=IOTESTTESTTEST",
+                 thing_name: _
+               }
+             } =
+               Devices.get_device_ws_endpoint(
+                 device,
+                 user,
+                 datetime: {{2020, 01, 01}, {12, 34, 56}}
+               )
+    end
+  end
+
+  test "get ws endpoint, error on get endpoint", %{device: device, user: user} do
+    use_cassette "aws/devices_get_ws_endpoint#2", custom: true do
+      new_config =
+        :cadet
+        |> Application.get_env(:remote_execution)
+        |> Keyword.delete(:endpoint_address)
+
+      Application.put_env(:cadet, :remote_execution, new_config)
+
+      assert {:error, _} = Devices.get_device_ws_endpoint(device.id, user)
+    end
+  end
+
+  test "get ws endpoint, error on assume role", %{device: device, user: user} do
+    use_cassette "aws/devices_get_ws_endpoint#3", custom: true do
+      assert {:error, _} = Devices.get_device_ws_endpoint(device.id, user)
+    end
+  end
+
+  defp assert_submaps_eq(expected, actual, fields) do
+    assert length(expected) == length(actual)
+
+    Enum.map(Enum.zip([expected, actual]), fn {e, a} -> assert_submap_eq(e, a, fields) end)
+  end
```

Comment:
This looks like it's essentially checking for structural equality - not sure if Elixir's `Map.equal?/2` does this. That aside, this seems generally useful enough to be moved into a file for test helpers.